src/com/xilinx/rapidwright/timing/delayestimator/DelayEstimatorBase.java

```diff
@@ -0,0 +1,461 @@
+    // Using enum as keys to simplify coding. Some types have empty arrays because those types will never be used.
+    // distArrays are cumulative and inclusive, ie., for a segment spanning x-y, d[y] is included in d of the segment.
+    protected Map<T.Orientation,Map<GroupDelayType, List<Short>>> distArrays;
+    protected int numCol;
+    protected int numRow;
+    
+    // These data are sourced from TimingModel.
+    // TODO: Consider to move these to TimingModel.
+    protected Map<T.Orientation,Map<GroupDelayType, Float>> K0;
+    protected Map<T.Orientation,Map<GroupDelayType, Float>> K1;
+    protected Map<T.Orientation,Map<GroupDelayType, Float>> K2;
+    protected Map<T.Orientation,Map<GroupDelayType, Short>> L;
+    protected Map<String, Short>  inputSitePinDelay;
+
+    protected T     ictInfo;
+    protected int   verbose;
+    protected transient Device device;
+    protected boolean useUTurnNodes;
+
+
+    /**
+     * Constructor from a device.
+     * Package scope to disable creating DelayEstimatorBase by a user.
+     * Create one using DelayEstimatorBuilder instead.
+     * @param device target device.
+     */
+    public DelayEstimatorBase(Device device, T ictInfo, boolean useUTurnNodes, int verbose) {
+        this.device  = device;
+        this.verbose = verbose;
+        this.ictInfo = ictInfo;
+        this.useUTurnNodes = useUTurnNodes;
+        TimingModel timingModel = new TimingModel(device);
+        timingModel.build();
+        buildDistanceArrays(timingModel);
+        loadInputSitePinDelay(timingModel);
+    }
+
+
+    /**
+     * Check if the node is a long node or not
+     * @param node  the node to be checked
+     * @return      true if the node is a long node
+     */
+    public static boolean isLong(Node node) {
+        return node.getIntentCode() == IntentCode.NODE_VLONG || node.getIntentCode() ==IntentCode.NODE_HLONG;
+    }
+
+
+    /**
+     * Return an extra delay if both parent and child are long node.
+     * @param child        a child node
+     * @param longParent   an indicator if the parent is a long node
+     * @return             the extra delay if any
+     */
+    public static short getExtraDelay(Node child, boolean longParent) {
+    	if(!longParent) return 0;
+    	
+    	IntentCode icChild = child.getIntentCode();
+    	if((icChild == IntentCode.NODE_VLONG) || (icChild == IntentCode.NODE_HLONG)) {
+    		return 45;
+    	}
+    	return 0;
+    }
+
+
+    public short getDelayOf(Node exitNode) {
+	    RoutingNode node = getTermInfo(exitNode);
+	    
+	    // Don't put this in calcTimingGroupDelay because it is called many times to estimate delay.
+	    if (node.ng == T.NodeGroupType.CLE_IN) {
+	    	return inputSitePinDelay.getOrDefault(exitNode.getWireName(), (short) 0);
+	    }
+	    
+	    return calcTimingGroupDelay(node.ng, node.begin(), node.end(), 0d);
+    }
+
+    /**
+     * Represent a routing resource in the delay graph
+     */
+    protected class RoutingNode {
```

Comment:
There is another inner class called `RoutingNode`. Do we need two?  Can we unify them?  I'd like to avoid inner classes.